File: apps/backend/src/rhesis/backend/app/schemas/telemetry.py
"""Backend telemetry schemas - imports from SDK."""

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field

# Import canonical schemas from SDK (source of truth)
from rhesis.sdk.telemetry.schemas import (
    AILLMAttributes,
    AIOperationType,
    AIToolAttributes,
    OTELSpan,
    OTELTraceBatch,
    SpanEvent,
    SpanKind,
    SpanLink,
    StatusCode,
    TraceIngestResponse,
)

from .comment import Comment
from .tag import TagRead

# Re-export SDK schemas for backward compatibility
__all__ = [
    "SpanKind",
    "StatusCode",
    "AIOperationType",
    "SpanEvent",
    "SpanLink",
    "AILLMAttributes",
    "AIToolAttributes",
    "OTELSpan",
    "OTELTraceBatch",
    "TraceIngestResponse",
    "OTELSpanResponse",
    "OTELSpanDB",
    "TraceSummary",
    "TraceListResponse",
    "SpanNode",
    "TraceDetailResponse",
    "TraceMetricsResponse",
]

# Alias for consistency with existing backend code
OTELSpanCreate = OTELSpan


# Backend-specific models (not in SDK)
class OTELSpanResponse(BaseModel):
    """Schema for returning a span from the database."""

    id: str
    trace_id: str
    span_id: str
    parent_span_id: Optional[str]
    project_id: str
    organization_id: str
    environment: str
    span_name: str
    span_kind: str
    start_time: datetime
    end_time: datetime
    duration_ms: float
    status_code: str
    status_message: Optional[str]
    attributes: Dict[str, Any]
    events: List[Dict[str, Any]]
    links: List[Dict[str, Any]]
    resource: Dict[str, Any]
    processed_at: Optional[datetime]
    enriched_data: Dict[str, Any]
    created_at: datetime
    updated_at: datetime

    # Tags and comments
    tags: Optional[List[TagRead]] = Field(
        default_factory=list, description="Tags associated with this trace"
    )
    comments: Optional[List[Comment]] = Field(
        default_factory=list, description="Comments associated with this trace"
    )

    model_config = ConfigDict(from_attributes=True)


class OTELSpanDB(BaseModel):
    """Database model with persistence fields."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    organization_id: str
    processed_at: Optional[datetime] = None
    enriched_data: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime
    span_data: OTELSpan


# Legacy alias (deprecated - use TraceIngestResponse)
TraceResponse = TraceIngestResponse


# Query response schemas
class TraceSummary(BaseModel):
    """Summary of a trace for list view."""

    trace_id: str
    project_id: str
    environment: str
    start_time: datetime
    duration_ms: float
    span_count: int
    root_operation: str
    status_code: str
    total_tokens: Optional[int] = None
    total_cost_usd: Optional[float] = None
    total_cost_eur: Optional[float] = None
    has_errors: bool

    # Test execution context (optional - only present for test execution traces)
    test_run_id: Optional[str] = Field(
        default=None, description="Test run ID if this trace is from a test execution"
    )
    test_result_id: Optional[str] = Field(
        default=None, description="Test result ID if this trace is from a test execution"
    )
    test_id: Optional[str] = Field(
        default=None, description="Test ID if this trace is from a test execution"
    )

    # Tags and comments count for summary view
    tags_count: Optional[int] = Field(
        default=0, description="Number of tags associated with this trace"
    )
    comments_count: Optional[int] = Field(
        default=0, description="Number of comments associated with this trace"
    )

    model_config = ConfigDict(from_attributes=True)


class TraceListResponse(BaseModel):
    """Response for list traces endpoint."""

    traces: List[TraceSummary]
    total: int = Field(..., description="Total traces matching filters")
    limit: int = Field(..., description="Results per page")
    offset: int = Field(..., description="Pagination offset")


class SpanNode(BaseModel):
    """Span node in trace tree."""

    span_id: str
    span_name: str
    span_kind: str
    start_time: datetime
    end_time: datetime
    duration_ms: float
    status_code: str
    status_message: Optional[str]
    attributes: Dict[str, Any]
    events: List[Dict[str, Any]]
    children: List["SpanNode"] = Field(default_factory=list)

    # Tags and comments
    tags: Optional[List[TagRead]] = Field(
        default_factory=list, description="Tags associated with this span"
    )
    comments: Optional[List[Comment]] = Field(
        default_factory=list, description="Comments associated with this span"
    )


class TraceDetailResponse(BaseModel):
    """Detailed trace with full span tree."""

    trace_id: str
    project_id: str
    environment: str
    start_time: datetime
    end_time: datetime
    duration_ms: float
    span_count: int
    error_count: int
    total_tokens: int
    total_cost_usd: float
    root_spans: List[SpanNode]


class TraceMetricsResponse(BaseModel):
    """Aggregated metrics for traces."""

    total_traces: int
    total_spans: int
    total_tokens: int
    total_cost_usd: float
    error_rate: float
    avg_duration_ms: float
    p50_duration_ms: float
    p95_duration_ms: float
    p99_duration_ms: float
    operation_breakdown: Dict[str, int]  # operation_type -> count

File: apps/frontend/src/utils/api-client/interfaces/telemetry.ts
/**
 * Telemetry API interfaces matching backend schemas
 * from apps/backend/src/rhesis/backend/app/schemas/telemetry.py
 */

import type { Execution, Verdict } from '@/constants/outcomes';
import type { AnnotationSummaryEntry } from './annotation';

/** Span event with known structure from OpenTelemetry */
export interface SpanEvent {
  name: string;
  timestamp?: string;
  attributes?: Record<string, string | number | boolean>;
  [key: string]: unknown;
}

/**
 * Span node in trace tree with hierarchical children
 */
export interface SpanNode {
  id?: string;
  span_id: string;
  span_name: string;
  span_kind: string;
  start_time: string;
  end_time: string;
  duration_ms: number;
  status_code: string;
  status_message?: string;
  attributes: Record<string, string | number | boolean>;
  /** Priced by enrichment; absent until a trace has been enriched. */
  cost_usd?: number | null;
  model_name?: string | null;
  events: SpanEvent[];
  children: SpanNode[];
  trace_metrics?: Record<string, unknown>;
  execution: Execution;
  verdict: Verdict | null;
  last_annotation?: AnnotationSummaryEntry;
  matches_annotation?: boolean;
  annotation_summary?: Record<string, AnnotationSummaryEntry>;
}

/**
 * Trace summary for list view
 */
export interface TraceSummary {
  trace_id: string;
  project_id: string;
  environment: string;
  conversation_id?: string;
  conversation_input?: string;
  start_time: string;
  duration_ms: number;
  span_count: number;
  root_operation: string;

  // Token usage and cost. Summed over the trace's llm.invoke spans, so a trace with
  // no LLM spans omits them rather than reporting zero.
  total_tokens?: number | null;
  total_input_tokens?: number | null;
  total_output_tokens?: number | null;
  total_cost_usd?: number | null;
  total_input_cost_usd?: number | null;
  total_output_cost_usd?: number | null;

  // What produced those tokens. `models` is empty only when nothing was priced, which
  // is how a caller tells that apart from a trace that genuinely cost nothing.
  models?: string[];
  providers?: string[];

  // Endpoint information (optional)
  endpoint_id?: string;
  endpoint_name?: string;

  // Trace metrics evaluation. execution/verdict are the source of truth
  // (see constants/outcomes.ts); trace_metrics_status is the legacy display
  // name kept alongside them.
  trace_metrics_status?: TraceMetricsStatus;
  execution: Execution;
  verdict: Verdict | null;

  // Human annotations
  has_annotations?: boolean;
  last_annotation?: AnnotationSummaryEntry;
  matches_annotation?: boolean;
}

/**
 * Detailed trace response with full span tree
 */
export interface TraceDetailResponse {
  trace_id: string;
  environment: string;
  conversation_id?: string;
  duration_ms: number;
  span_count: number;
  error_count: number;

  // Token usage and cost across the trace's llm.invoke spans. The detail endpoint
  // has the whole span set, so unlike TraceSummary it always resolves the split.
  total_tokens: number;
  total_input_tokens: number;
  total_output_tokens: number;
  total_cost_usd: number;

  root_spans: SpanNode[];

  // Trace metrics evaluation. execution/verdict are the source of truth
  // (see constants/outcomes.ts); trace_metrics_status is the legacy display
  // name kept alongside them.
  trace_metrics_status?: TraceMetricsStatus;
  execution: Execution;
  verdict: Verdict | null;

  // Human annotations
  last_annotation?: AnnotationSummaryEntry;
  matches_annotation?: boolean;
  annotation_summary?: Record<string, AnnotationSummaryEntry>;

  // Related entities (optional - populated via relationships)
  project?: {
    id: string;
    name: string;
  };
  endpoint?: {
    id: string;
    name: string;
  };
  test_run?: {
    id: string;
    name?: string;
    nano_id?: string;
  };
  test_result?: {
    id: string;
  };
  test?: {
    id: string;
    nano_id?: string;
  };
}

/**
 * Paginated list response
 */
export interface TraceListResponse {
  traces: TraceSummary[];
  total: number;
}

/**
 * Trace source filter enum
 */
export type TraceSource = 'all' | 'test' | 'operation';

/**
 * Trace type filter enum (single-turn vs multi-turn)
 */
export type TraceType = 'all' | 'Single-Turn' | 'Multi-Turn';

/**
 * Evaluation status for trace metrics (Pass/Fail/Error)
 */
export type TraceMetricsStatus = 'Pass' | 'Fail' | 'Error';

export const TRACE_METRICS_STATUS = {
  PASS: 'Pass' as const,
  FAIL: 'Fail' as const,
  ERROR: 'Error' as const,
};

/**
 * Query parameters for list endpoint
 */
export interface TraceQueryParams {
  project_id?: string; // Optional - shows all projects if not specified
  environment?: string;
  /** Case-insensitive search across trace ID, operations, endpoint metadata, conversation text */
  search?: string;
  start_time_after?: string;
  start_time_before?: string;
  test_run_id?: string;
  test_result_id?: string;
  test_id?: string;
  endpoint_id?: string; // Filter by endpoint ID
  trace_source?: TraceSource; // Filter by trace source (all/test/operation)
  trace_type?: TraceType; // Filter by trace type (all/Single-Turn/Multi-Turn)
  trace_metrics_status?: TraceMetricsStatus;
  /** Repeatable: a trace matches if any of its priced calls used one of these. */
  provider?: string[];
  limit?: number;
  offset?: number;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

/**
 * Aggregated metrics response
 */
export interface TraceMetricsResponse {
  total_traces: number;
  total_spans: number;
  total_tokens: number;
  total_cost_usd: number;
  error_rate: number;
  avg_duration_ms: number;
  p50_duration_ms: number;
  p95_duration_ms: number;
  p99_duration_ms: number;
  operation_breakdown: Record<string, number>;
}
